### ISSUE TITLE
Non-existent constructor doesn't error before redundant

### ISSUE THREAD

**Issue by Alizter** [coq/coq#17195]
Here is a erroneous match:
```coq
Inductive A := case_exists : A.

Definition foo x := match x with
| case_doesnt_exist => True
| case_exists => True
end.
```

Coq complains that:
```
Pattern "case_exists" is redundant in this clause.
```

But really, rather than checking case_exists is redundant first, it should have caught that `case_doesnt_exist` doesn't exist.

**Comment by SkySkimmer** [comment]
But non existent constructor is a warning (unused-pattern-matching-variable), I'm not sure it would be useful to try to move its check earlier.

**Comment by Alizter** [comment]
Would it make sense to have a syntax like ```
`foo
```
To enforce that the pattern be a constructor?

I suppose I got confused about the original pattern matching due to the behaviour in OCaml.

**Comment by herbelin** [comment]
> But non existent constructor is a warning (unused-pattern-matching-variable), I'm not sure it would be useful to try to move its check earlier.

Indeed. Not clear how to do it though (currently, the `case_exists` branch is dropped early and the warning is lost).

Otherwise, we could try to detect the above situation, but we need a criterion which works also in nested situations such as:
```coq
match x, y with
| true, false_mispelled => True
| z, true => True
end
```
where `false_mispelled` makes sense as a variable for the `false, false` case.

> Would it make sense to have a syntax like
> 
> ```coq
> `foo
> ```

Why not, or even `'foo`, consistently with `fun 'constructor => ...`. Then, we could apply a translator...

**Comment by herbelin** [comment]
An alternative to quoting foo could be to generalize the convention that unbound named pattern variables need to be prefixed with an underscore. Currently, this is done in (not necessary toplevel) situations such as
```coq
match x with
| C => True
| x => True
end
```
but not in situations as above, i.e. like
```coq
match x with
| x => True
| C => True
end
```
nor in
```coq
match x with
| x => True
end
```
Expecting `x` to be `_x` in the latter would require a lot of changes in existing code, but expecting it in the second configuration is more realistic (as @SkySkimmer implicitly suggests) (though a bit hackish to implement at first view).

**Comment by SkySkimmer** [comment]
>generalize the convention that unbound named pattern variables need to be prefixed with an underscore

unbound or unused?

**Comment by herbelin** [comment]
yes, unused would have been a better word here.
